## Differential Expression

The table below summarizes the results of the differential tests with the name of the comparison, the model (if applicable), the test (either the coefficient or contrast tested), the number of differential loci in case and control, and the number of loci tested but not statistically significant.

The following sections provide more detailed information about the differential tests, including a summary table of the top differential loci (annotated to genomic features if applicable), a volcano plot, a plot summarizing the annotations of the differential loci compared to the background of all tested loci, and a filtering of the differential loci according to genes of interest (if applicable).

```{r include = FALSE}
    info = snakemake@params[['diffex_model_info']]
    names(info) = gsub('_', ' ', names(info)) # Remove _ from model names
    info_tbl = do.call(rbind, lapply(info, data.frame))
    info_tbl[,'factor_name'] = NULL # factor_name doesn't need to be put in report, used internally by Rscripts
    info_tbl = rownames_to_column(info_tbl, var = 'Model Name')
    colnames(info_tbl) = c('Model Name', 'Model', 'Linear fold-change', 'FDR')
    info_kable = kable(info_tbl, row.names = TRUE, format.args = list(big.mark = ',', scientific = FALSE)) %>%
      kable_styling(bootstrap_options = c('striped', 'condensed'))
```

```{r echo = FALSE, results = 'asis', message = FALSE}
    cat(info_kable)
```

```{r echo = FALSE, results = 'asis', message = FALSE}
    comparisons = snakemake@params[['contrasts']]
    names(comparisons) = gsub('_', ' ', names(info)) # Remove _'s from model names - consistent with info list above
    sample_phenotypes = snakemake@params[['sample_phenotypes']]
    phenotypes = names(sample_phenotypes)

    for(model_name in names(comparisons)) {
        #Get the samples for case and control
        factor_name = info[[model_name]][['factor_name']]
        #Function to get case or control samples (concatenated strings)
        #using the sample phenotype list with the comparison string
        get_case_ctrl_samples = function(sample_phenotypes, factor, comp_str, case = FALSE, ctrl = FALSE){
          comp.split = unlist(strsplit(comp_str, "_v_"))
          case_name = comp.split[1] ; ctrl_name = comp.split[2]
          if(case == TRUE) {
            return(paste(sample_phenotypes[[factor]][[case_name]], collapse='<br>'))
          } else if (ctrl == TRUE) {
            return(paste(sample_phenotypes[[factor]][[ctrl_name]], collapse='<br>'))
          }
        }

        comps_list = comparisons[[model_name]]
        case_smpls = sapply(X = comps_list, FUN = function(x){
            return(get_case_ctrl_samples(sample_phenotypes, factor_name, x, case = TRUE))
        })
        ctrl_smpls = sapply(X = comps_list, FUN = function(x){
            return(get_case_ctrl_samples(sample_phenotypes, factor_name, x, ctrl = TRUE))
        })

        comparison_tbl = data.frame(cbind(comps_list, case_smpls, ctrl_smpls))
        colnames(comparison_tbl) = c('Comparison', 'Case Samples', 'Ctrl Samples')

        comparison_kable = kable(comparison_tbl, format = 'html', row.names = FALSE, escape = FALSE, format.args = list(big.mark = ',', scientific = FALSE)) %>%
          kable_styling(bootstrap_options = c('striped', 'condensed')) %>%
          scroll_box(width = '100%')

        filter_sig = info[[model_name]][['adjustedPValue']]
        filter_fc = info[[model_name]][['linear_fold_change']]
        cat(sprintf('#### Comparisons for %s\n\n', model_name))
        cat(comparison_kable)
        cat(sprintf('In the volcano plots below, each point is a tested locus. A gene is considered to be differentially expressed if FDR < %s and | linear fold-change | > %s. Points are colored according to the direction of differential expression with the total number of loci indicated in the legend. The vertical lines correspond to the fold-change cutoff. \n\n', filter_sig, filter_fc))
        for(comparison in comps_list){
          model_name_for_files = gsub(' ', '_', model_name) # filepaths have underscores instead of spaces
          volcano_file = sprintf(diffex_volcano_file, diffex_dir, model_name_for_files, comparison)
          cat(sprintf('![**Figure:** Volcano plot for %s.](%s) \n\n', comparison, volcano_file))
          annot_file = sprintf(diffex_annot_file, diffex_dir, model_name_for_files, comparison)
          annot_tbl = read.delim(annot_file)

          annot_tbl = annot_tbl[1:100,] %>%
              mutate(entrezgene_id = cell_spec(
                x = entrezgene_id,
                format = "html",
                link = paste0('https://www.ncbi.nlm.nih.gov/gene/?term=', entrezgene_id)
              ))
          annot_kable = kable(annot_tbl, "html", escape = FALSE, row.names = FALSE, format.args = list(big.mark = ',', scientific = FALSE)) %>%
              kable_styling(bootstrap_options = c('striped', 'condensed')) %>%
              scroll_box(width = '100%', height = '400px')
          cat(sprintf('Top differentially expressed genes for %s \n\n', comparison))
          cat(annot_kable)
        }
    }
```