## QC and sequence alignment

The reads were trimmed using Cutadapt v`r versions[['cutadapt']][['cutadapt']]`. Reads were mapped to the reference genome `r config[['genome']]` (ENSEMBL), using STAR v`r versions[['rsem_star']][['star']]` [@Dobin2013] and assigned count estimates to genes with RSEM v`r versions[['rsem_star']][['rsem']]` [@Li2011]. Alignment options followed ENCODE standards for RNA-seq [@Dobin:STARmanual]. FastQC [@Andrews2010] (v`r versions [['fastqc']][['fastqc']]`) was run on .bam files in a post-alignment step, including both aligned and un-aligned reads, to ensure data quality. 
