## QC and sequence alignment

The reads were trimmed using Cutadapt v`r versions[['cutadapt']][['cutadapt']]` [@Martin2011]. The reads were evaluated with FastQC [@Andrews2010] (v`r versions [['fastqc']][['fastqc']]`) to determine quality of the data. Reads were mapped to the reference genome `r config[['genome']]` (ENSEMBL), using STAR v`r versions[['rsem_star']][['star']]` [@Dobin2013] and assigned count estimates to genes with RSEM v`r versions[['rsem_star']][['rsem']]` [@Li2011]. Alignment options followed ENCODE standards for RNA-seq [@Dobin:STARmanual]. QC metrics from several different steps in the pipeline were aggregated by multiQC v`r versions[['multiqc']][['multiqc']]` [@Ewels2016]. 
