## Pipeline

We use Snakemake [@Koster2012] to manage the bioinformatics workflow in a reproducible manner.

## QC and sequence alignment

At the UM Bioinformatics Core, we downloaded the reads files from the Sequencing Coreâ€™s storage and trimmed adapters using Cutadapt (v`r versions[['cutadapt']][['cutadapt']]`). FastQC [@Andrews2010] (v`r versions [['fastqc']][['fastqc']]`) was used to ensure the quality of data. Reads were mapped to the reference genome `r config[['genome']]` (ENSEMBL), using STAR [@Dobin2013] (v`r versions[['rsem_star']][['star']]`) and assigned count estimates to genes with RSEM [@Li2011] (v`r versions[['rsem_star']][['rsem']]`). Alignment options followed ENCODE standards for RNA-seq [@Dobin:STARmanual]. FastQC was used in an additional post-alignment step to ensure that only high-quality data were used for expression quantitation and differential expression.
