# Results

## QC

A detailed MultiQC report can be found at the following link:

```{r include = FALSE}
# Setup - generate relative link to multiqc html
cwd = sub("/ccmb/BioinfCore/ActiveProjects", "/nfs/med-bfx-activeprojects", getwd()) # TODO: when we remove /ccmb/ completely this will no longer be necessary, can just use getwd()
mqc_fullpath = file.path(project_dir, snakemake@input[['multiqc_html']])
realpath_args = c('--relative-to', cwd, mqc_fullpath)

multiqc_link_rel = xml2::url_relative(mqc_fullpath, file.path(cwd, '..'))
```
  
### [Alignment Results - MultiQC report](`r multiqc_link_rel`)

### Alignment Results - Summary & Selected Figures

A summary of alignment statistics can be seen in the table below.


```{r include = FALSE}
# Read in multiqc tab-separated files
multiqc_gen_file = file.path(project_dir, snakemake@input[['multiqc_gen_stats']])
multiqc_star_file = file.path(project_dir, snakemake@input[['multiqc_star']])

multiqc_gen = read.table(multiqc_gen_file, header = TRUE, sep='\t', quote = "", stringsAsFactors = FALSE)
multiqc_star = read.table(multiqc_star_file, header = TRUE, sep='\t', quote = "", stringsAsFactors = FALSE)

# Subset STAR table to desired properties
star_keep_cols = c('Sample', 'total_reads', 'uniquely_mapped', 'uniquely_mapped_percent', 'avg_mapped_read_length', 'multimapped', 'multimapped_percent')
multiqc_star_subset = multiqc_star[,star_keep_cols]

# Subset general table to desired properties (just fastqc aligned percent-dup for now)
fastqc_aligned_pcdups = grep("^FastQC.*aligned.*percent_duplicates$", colnames(multiqc_gen), value = TRUE)
multiqc_gen_subset = multiqc_gen[,c('Sample', fastqc_aligned_pcdups)] #Just grab Sample and aligned dup percentage from general stats

# Function to trim e.g. foo.bar.baz to baz (for simplifying column names)
trim_to_suffix = function(fullname) {return(sub(".*\\.", "", fullname))} # Trim to final suffix (all else captured by greedy . match)

# Clean column names
colnames(multiqc_gen_subset) = trim_to_suffix(colnames(multiqc_gen_subset))

# Remove rows with NA (in this case, everything except fastqc_aligned data rows)
multiqc_gen_subset = multiqc_gen_subset[complete.cases(multiqc_gen_subset),]

# Combine subset of STAR results table with subset of general results table
multiqc_report_df = merge(multiqc_star_subset, multiqc_gen_subset, by = 'Sample', all = TRUE)

# Transform the column names
colnames(multiqc_report_df) = gsub("_", " ", colnames(multiqc_report_df))

multiqc_kable = kable(multiqc_report_df, row.names = FALSE, format.args = list(big.mark = ',', scientific = FALSE)) %>%
  kable_styling(bootstrap_options = c('striped', 'condensed')) %>%
  scroll_box(width = '100%', height = '600px')
```

```{r echo = FALSE, results = 'asis', message = FALSE}
cat(multiqc_kable)
```

```{r eval = include_diffex, echo = FALSE, results = 'asis', message = FALSE}
cat('### Boxplots\n\nThe following boxplots show the distribution of gene expression counts in the different samples.\n\n')
for(factor_cov in snakemake@params[['phenotypes']]) {
  cat(sprintf('\n\n#### %s \n\n', factor_cov))
  # Include 'raw' if desired
  for(value in c('raw', 'rlog')) {
    file = sprintf(qc_boxplot_file, diffex_dir, factor_cov, value)
    if(value == 'rlog'){
      value = 'rlog normalized' #This is nicer for the caption
    }
    cat(sprintf('![**Figure:** Box plot of %s gene expression colored by %s.](%s) \n\n', value, factor_cov, file))
  }
}
```

```{r eval = include_diffex, echo = FALSE, results = 'asis', message = FALSE}
cat('### Clustering\n\nClustering is a method to visualize the similarity and difference among samples. Unsupervised clustering using complete linkage of the Euclidian distances of the gene expression profiles of pairwise samples is shown below. [REPLACE: Anything of note? Mention it here.]\n\n')
for(factor_cov in snakemake@params[['phenotypes']]) {
  cat(sprintf('\n\n#### %s \n\n', factor_cov))
  file = sprintf(qc_heatmap_file, diffex_dir, factor_cov)
  cat(sprintf('![**Figure:** Unsupervised clustering of %s expression values colored by %s.](%s) \n\n', value, factor_cov, file))
}
```

```{r eval = include_diffex, echo = FALSE, results = 'asis', message = FALSE}
cat('### PCA\n\nPrincipal component analysis (PCA) is another method to visualize the similarity and difference among samples. PCA was applied to the normalized gene expression profiles of the samples. [REPLACE: Anything of note? Mention it here.]\n\n')
for(factor_cov in snakemake@params[['phenotypes']]) {
  cat(sprintf('\n\n#### %s \n\n', factor_cov))
  # Include 'raw' if desired
  for(value in c('top100', 'top500')) {
    pca_12_png = sprintf(qc_pca_file, diffex_dir, factor_cov, value)
    cat(sprintf('![**Figure:** PCA plot (PC1 vs PC2) of %s expression values colored by %s.](%s) \n\n', value, factor_cov, pca_12_png))
  }
}
```