## QC and sequence alignment

At the UM Bioinformatics Core, we downloaded fastq files from the Sequencing Coreâ€™s storage and trimmed the reads using Cutadapt v`r versions[['cutadapt']][['cutadapt']]`. FastQC [@Andrews2010] (v`r versions [['fastqc']][['fastqc']]`) was used to ensure the quality of data. Reads were mapped to the reference genome `r config[['genome']]` (ENSEMBL), using STAR v`r versions[['rsem_star']][['star']]` [@Dobin2013] and assigned count estimates to genes with RSEM v`r versions[['rsem_star']][['rsem']]` [@Li2011]. Alignment options followed ENCODE standards for RNA-seq [@Dobin:STARmanual]. FastQC was used in an additional post-alignment step to ensure that only high-quality data were used for expression quantitation and differential expression.
