---
title: "Methods"
output:
  pdf_document:
    fig_caption: true
header_includes:
  \usepackage{float}
  \floatplacement{figure}{H}
bibliography: references_WAT.bib
csl: bioinformatics.csl
---

<!--- Allow the page to be wider --->
<style>
  body .main-container {
    max-width: 1200px;
  }
</style>

```{r include = FALSE}
################################################################################
# Load Libraries
library(yaml)


################################################################################
# Set things up you'll need downstream

config = snakemake@config
setwd(project_dir)

# Read the software environment version yaml
versions = yaml.load_file(snakemake@input[['versions']])
```


<!---
Methods
--->

A) The reads were trimmed using Cutadapt v`r versions[['cutadapt']][['cutadapt']]` [@Martin2011].
B) FastQC v`r versions [['fastqc']][['fastqc']]` was used to ensure the quality of data [@Andrews2010].
C) Fastq Screen v`r versions[['fastq_screen']][['fastq-screen']]` was used to screen for various types of contamination [@Wingett2018].
D) Reads were mapped to the reference genome `r config[['genome']]` (ENSEMBL), using STAR v`r versions[['rsem_star']][['star']]` [@Dobin2013] and assigned count estimates to genes with RSEM v`r versions[['rsem_star']][['rsem']]` [@Li2011]. Alignment options followed ENCODE standards for RNA-seq.
E) Multiqc v`r versions[['multiqc']][['multiqc']]` compiled the results from several of these tools and provided a detailed and comprehensive quality control report [@Ewels2016]. 

\  
\  

```{r echo = FALSE, results = 'asis', message = FALSE}
cat(sprintf('![**Figure:** Methods overview](%s){width=400px} \n\n', methods_fig))
```

\  
\  
