---
title: "`r project_name` Project Report"
author: "`r doc_author`"
date: "`r Sys.Date()`"
output:
        html_document:
            theme: readable
            toc: true
            toc_depth: 4
            toc_float: true
            number_sections: true
            fig_caption: true
            keep_md: true
bibliography: references_WAT.bib
csl: bioinformatics.csl
---

<!---
Can change this either in the .md or .Rmd document...

For docx output to get edits, change output field above to:
        word_document:
            toc: true
            toc_depth: 3
            fig_caption: true
            keep_md: true

Default HTML output:

        html_document:
            toc: true
            toc_depth: 3
            toc_float: true
            keep_md: true
--->

<!--- Allow the page to be wider --->
<style>
    body .main-container {
        max-width: 1200px;
    }
</style>

```{r include = FALSE}

################################################################################
# Set things up you'll need downstream

knitr::opts_knit$set(root.dir = project_dir)

################################################################################
# Flags for showing additional optional sections

show_background_biology = TRUE # TODO: These could alternatively be moved out to config
show_background_samples = TRUE
show_follow_up = config[['report_info']][['include_follow_up']] # note this is in config, others set above

```

*Contact:*

* `r analyst_name` <`r analyst_email`>

<!---
Background
--->


# Background

```{r background_biology_section, child = file.path(REPORT_SRC_DIR, 'child_background_biology.Rmd'), eval = show_background_biology}
```

```{r background_samples_section, child = file.path(REPORT_SRC_DIR, 'child_background_samples.Rmd'), eval = show_background_samples}
```

<!---
Methods
--->


# Methods

## Sample preparation & sequencing

```{r echo = FALSE, results = 'asis', message = FALSE}
    cat(config[['report_info']][['prep_description']])
```

## Differential Expression

Data were pre-filtered to remove genes with 0 counts in all samples. Differential gene expression analysis was performed using DESeq2 [@Love2014], using a negative binomial generalized linear model (thresholds: linear fold change >1.5 or <-1.5, Benjamini-Hochberg FDR (Padj) <0.05). Plots were generated using variations of DESeq2 plotting functions and other packages with R version `r sw_versions[['WAT_diffex']][['r-base']]`. Annotation data from ENSEMBL `r config[['ensembl_version']]` was used, and genes were additionally annotated with Entrez GeneIDs and text descriptions. Functional analysis, including candidate pathways activated or inhibited in comparison(s) and GO-term enrichments [@Mi2019], was performed using iPathway Guide (Advaita) [@ipathwayguide], [@Draghici2007].

<!---
Results
--->

# Results
<!---
results QC diffex
--->
```{r echo = FALSE, results = 'asis', message = FALSE}

cat('## Boxplots\n\nThe following boxplots show the distribution of gene expression counts in the different samples.\n\n')
for(factor_cov in phenotypes) {
  cat(sprintf('\n\n### %s \n\n', factor_cov))
  # Include 'raw' if desired
  for(value in c('raw', 'rlog')) {
    file = normalizePath(sprintf(qc_boxplot_file, DIFFEX_DIR, factor_cov, value))
    if(value == 'rlog'){
      value = 'rlog normalized' #This is nicer for the caption
    }
    cat(sprintf('![**Figure:** Box plot of %s gene expression colored by %s.](%s) \n\n', value, factor_cov, file))
  }
}
```

```{r echo = FALSE, results = 'asis', message = FALSE}
cat('## Clustering\n\nClustering is a method to visualize the similarity and difference among samples. Unsupervised clustering using complete linkage of the Euclidian distances of the gene expression profiles of pairwise samples is shown below.\n\n')

file = normalizePath(sprintf(qc_heatmap_file, DIFFEX_DIR))
cat(sprintf('![**Figure:** Unsupervised clustering of %s expression values colored by %s.](%s) \n\n', value, paste(phenotypes, collapse = ', '), file))
```

```{r echo = FALSE, results = 'asis', message = FALSE}
cat('## PCA\n\nPrincipal component analysis (PCA) is another method to visualize the similarity and difference among samples. PCA was applied to the normalized gene expression profiles of the samples.\n\n')
for(factor_cov in phenotypes) {
  cat(sprintf('\n\n### %s \n\n', factor_cov))
  # Include 'raw' if desired
  for(value in c('top 100', 'top 500')) {
    val_nospace = sub(" ", "", value)
    pca_12_png = normalizePath(sprintf(qc_pca_file, DIFFEX_DIR, factor_cov, val_nospace))
    cat(sprintf('![**Figure:** PCA plot (PC1 vs PC2) of %s variable genes colored by %s.](%s) \n\n', value, factor_cov, pca_12_png))
  }
}
```

<!---
Main results diffex
--->
## Differential Expression

The tables below summarize the setup of the differential tests with the name of the comparison, the model, and the threshold values, as well as the samples included in each comparison and the number of differentially expressed genes.

The sections that follow provide more detailed information about the differential tests, including a summary table of the top differential loci (annotated to genomic features if applicable), and a volcano plot for each comparison.

```{r include = FALSE}
    info_df$model_name = gsub("_", " ", info_df$model_name)
    colname_key = c(model_name='Model Name', model='Model', linear_fold_change='Linear fold-change', adjustedPValue='FDR')
    colnames(info_df) = colname_key[colnames(info_df)]  # Rename the columns that are present, according to the key above
    info_kable = kable(info_df, format.args = list(big.mark = ',', scientific = FALSE)) %>%
      kable_styling(bootstrap_options = c('striped', 'condensed'))
```

```{r echo = FALSE, results = 'asis', message = FALSE}
    cat(info_kable)
```

```{r echo = FALSE, results = 'asis', message = FALSE}

    #Function to get case or control samples (concatenated strings)
    #using the samplesheet and each comparison string
    get_case_ctrl_samples = function(pdata, factor, comp_str, selection){
      comp.split = unlist(strsplit(comp_str, "_v_"))
      case_name = comp.split[1] ; ctrl_name = comp.split[2]
      if(selection == 'case'){
        samples_str = pdata[pdata[[factor]] == case_name, 'sample']
        samples_str = paste(samples_str, collapse='<br>') # Collapse on <br> to create multi-line entries in the table
      } else if(selection == 'ctrl'){
        samples_str = pdata[pdata[[factor]] == ctrl_name, 'sample']
        samples_str = paste(samples_str, collapse='<br>')
      }
      return(samples_str)
    }

    for(model_name in model_names) {
        #Get the samples for case and control
        factor_name = config[['diffex']][[model_name]][['DESeq2']][['factor_name']]

        if(!is.null(pdata_subset_list[[model_name]])){
          pdata_case_ctrl = pdata_subset_list[[model_name]]
        } else{
          pdata_case_ctrl = pdata_full
        }

        model_match_str = sub("model_", "", model_name)
        comps_list = unname(unlist(summary_df[summary_df$model_name == model_match_str, 'comparison']))

        de_counts = summary_df[summary_df$model_name == model_match_str, c("comparison", "count_diff_expressed")]

        filter_sig = config[['diffex']][[model_name]][['adjustedPValue']]
        filter_fc = config[['diffex']][[model_name]][['linear_fold_change']]

        # If comps_list is a single value 'results', this indicates there are no comparisons, so no comparison table nor volcanoplot
        if(comps_list != 'results'){
          case_smpls = sapply(X = comps_list, FUN = function(x){
              return(get_case_ctrl_samples(pdata_case_ctrl, factor_name, x, selection = 'case'))
          })
          ctrl_smpls = sapply(X = comps_list, FUN = function(x){
              return(get_case_ctrl_samples(pdata_case_ctrl, factor_name, x, selection = 'ctrl'))
          })

          case_ctrl_df = data.frame(cbind(comps_list, case_smpls, ctrl_smpls))

          comparison_tbl = merge(case_ctrl_df, de_counts, by.x="comps_list", by.y="comparison")

          colnames(comparison_tbl) = c('Comparison', 'Case Samples', 'Ctrl Samples', 'Diff. Exp.<br>Count')

          comparison_kable = kable(comparison_tbl, format = 'html', row.names = FALSE, escape = FALSE, format.args = list(big.mark = ',', scientific = FALSE)) %>%
            kable_styling(bootstrap_options = c('striped', 'condensed')) %>%
            scroll_box(width = '100%')

          cat(sprintf('### Comparisons for %s\n\n', gsub('_', ' ', model_name)))
          cat(comparison_kable)
          cat(sprintf('In the volcano plots below, each point is a tested locus. A gene is considered to be differentially expressed if FDR < %s and | linear fold-change | > %s. Points are colored according to the direction of differential expression with the total number of loci indicated in the legend. The vertical lines correspond to the fold-change cutoff. \n\n', filter_sig, filter_fc))
          for(comparison in comps_list){
            volcano_file = normalizePath(sprintf(diffex_volcano_file, DIFFEX_DIR, model_name, comparison))
            cat(sprintf('![**Figure:** Volcano plot for %s.](%s) \n\n', comparison, volcano_file))
          }
        }
        if(is.null(filter_fc)){
          cols_to_display = c('external_gene_name', 'entrezgene_id', 'description', 'padj')
        } else {
          cols_to_display = c('external_gene_name', 'entrezgene_id', 'description', 'log2FoldChange', 'padj')
        }
        for(comparison in comps_list) {
          annot_file = normalizePath(sprintf(diffex_annot_file, DIFFEX_DIR, model_name, comparison))
          # "." is used as NA value in the annotated DESeq2 result
          annot_tbl = read_delim(annot_file, delim="\t", na = c("", ".")) %>%
              slice_head(n = 200) %>%
              filter(Call == "YES") %>%
              dplyr::select(all_of(cols_to_display)) %>%
              mutate(padj = ifelse(padj < 0.001, format(padj, scientific=TRUE, digits=2), round(padj, digits=3))) %>%
              mutate(description = sub(" \\[Source:.*?\\]", "", description)) %>%
              mutate(entrezgene_id = ifelse(is.na(entrezgene_id), NA, cell_spec(
                  x = entrezgene_id,
                  format = "html",
                  link = paste0('https://www.ncbi.nlm.nih.gov/gene/?term=', entrezgene_id)
              )))

          cols_rename_key = c(external_gene_name='Gene', entrezgene_id='entrezgene_id', description='description', log2FoldChange='log2 (FC)', padj='q-val')
          colnames(annot_tbl) = cols_rename_key[colnames(annot_tbl)]  # Rename the columns that are present, according to the key above

          annot_kable = kable(annot_tbl, "html", digits = 3, escape = FALSE, row.names = FALSE, format.args = list(big.mark = ',', scientific = FALSE)) %>%
              kable_styling(bootstrap_options = c('striped', 'condensed')) %>%
              scroll_box(width = '100%', height = '400px')
          count_de = de_counts$count_diff_expressed[de_counts$comparison == comparison]
          if(comparison == 'results'){
            cat(sprintf('\n\n**Table:** Results\n\n%s differentially expressed genes. View limited to maximum of 200 rows in html report.', count_de))
          } else {
            cat(sprintf('\n\n**Table:** Results for %s \n\n%s differentially expressed genes. View limited to maximum of 200 rows in html report.', comparison, count_de))
          }
          cat(annot_kable)
        }
    }
```

```{r follow_up, child = file.path(REPORT_SRC_DIR, 'child_follow_up.Rmd'), eval = show_follow_up}
```


# Acknowledgement for Publication

`r acknowledgement_text`

<!---
Appendices
-->

# Appendices

## Differential Expression Files

### DESeq2 Counts

These are counts extracted from the DESeq2 object after filtering / normalization.

`counts/deseq2_raw_counts.txt`

raw_counts are the counts used in DESeq2, exactly as they are in the DESeq2 object (after filtering low/nonexpressed genes).

`counts/deseq2_depth_normalized_counts.txt`

These are the counts from above normalized to the sequencing depth by DESeq2.

`counts/deseq2_rlog_normalized_counts.txt`

Same as above, but normalized using the DESeq2's regularized log transformation to normalize.

`counts/deseq2_rlog_normalized_counts.annot.txt`

Rlog normalized counts as described above, with additional annotation columns

### Count-Level QC

These use the count data in various plots to illustrate the count-level QC. The plots are color-labeled based on phonotype / characteristic information from the experimental design. The versions with different color labels will each reside in their own folder.

`plots_labeled_by_pheno/{phenotype}/BoxPlot_raw.{pdf,png}`

BoxPlot of count data for each sample in the experiment.

`plots_labeled_by_pheno/{phenotype}/BoxPlot_rlog.{pdf,png}`

Boxplot of count data for each sample after rlog transformation.

`plots_labeled_by_pheno/Heatmap_TopExp.{pdf,png}`

Heatmap of count data, with data from up to 500 of the most highly expressed genes. Rows represent genes, columns represent samples.

`plots_labeled_by_pheno/Heatmap_TopVar.{pdf,png}`

Heatmap of count data, with data from up to 500 of the most variably expressed genes (across the samples). Rows represent genes, columns represent samples.

`plots_labeled_by_pheno/{phenotype}/PCAplot_{pcs}_top{ngenes}.{pdf,png}`

PCA plots of count data. {pcs} represents the principal components plotted 1 & 2 or 2 & 3. Top {ngenes} represents the number of highly expressed genes used in the PCA.

`plots_labeled_by_pheno/SampleHeatmap.{pdf,png}`

Heatmap of the euclidian distance between samples, based on the aggregated count data.

`plots_labeled_by_pheno/{phenotype}/ScreePlot_top{ngenes}.{pdf,png}`

Scree plots of the count data. Top {ngenes} represents the number of highly expressed genes used in the plot.

### DESeq2 Differential Expression Results

`diffex_{model_name}/{comparison}.annot.txt`

Output from DESeq2 differential expression analysis for a particular comparison, in plain tab-separated format, with additional annotations from BiomaRt

`diffex_{model_name}/{comparison}.annot.xlsx`

Output from DESeq2 differential expression analysis for a particular comparison, in xlsx format, with additional annotations from BiomaRt and hyperlinks to NCBI.

`volcano_plots/VolcanoPlot_{comparison}.{pdf,png}`

Volcano plots summarizing differential expression data for a particular comparison.

### DESeq2 Summary

`summary/deseq2_summary.txt`

Summarized information from the different comparisons performed with DESeq2.

`summary/deseq2_summary.xlsx`

An Excel spreadsheet containing the same information as above, with an additional column of entrezgeneID hyperlinks when possible

# References
