---
title: "`r project_name` Project Report"
author: "`r doc_author`"
date: "`r Sys.Date()`"
output:
        html_document:
            theme: readable
            toc: true
            toc_depth: 4
            toc_float: true
            number_sections: true
            fig_caption: true
            keep_md: true
bibliography: references_WAT.bib
csl: bioinformatics.csl
---

<!---
Can change this either in the .md or .Rmd document...

For docx output to get edits, change output field above to:
        word_document:
            toc: true
            toc_depth: 3
            fig_caption: true
            keep_md: true

Default HTML output:

        html_document:
            toc: true
            toc_depth: 3
            toc_float: true
            keep_md: true
--->

<!--- Allow the page to be wider --->
<style>
    body .main-container {
        max-width: 1200px;
    }
</style>

```{r include = FALSE}
################################################################################
# Load Libraries
library(yaml)


################################################################################
# Set things up you'll need downstream

config = snakemake@config
setwd(project_dir)

# Read the software environment version yaml
versions = yaml.load_file(snakemake@input[['versions']])

diffex_info = snakemake@params[['diffex_model_info']]

# Depending on snakemake rule, include/exclude sections (diffex, alignment, QC)
if(snakemake@rule == 'report_align_diffex'){
    show_methods_qc_sequencing = TRUE
    include_diffex = TRUE
    show_results_qc = TRUE
    show_appendix_qc = TRUE
} else if(snakemake@rule == 'report_align_only') {
    show_methods_qc_sequencing = TRUE
    include_diffex = FALSE
    show_results_qc = TRUE
    show_appendix_qc = TRUE
} else if(snakemake@rule == 'report_from_counts') {
    show_methods_qc_sequencing = FALSE
    include_diffex = TRUE
    show_results_qc = FALSE
    show_appendix_qc = FALSE
}

################################################################################
# Flags for showing additional optional sections

show_background_biology = snakemake@params[['add_background']]
show_methods_wetlab = snakemake@params[['add_wetlab']]
custom_choices = unlist(strsplit(snakemake@params[['custom_sections']], ','))

show_background_samples = TRUE

if(include_diffex) {
    show_background_diffex = TRUE
    show_methods_diffex = TRUE
    show_results_diffex = TRUE
    show_appendix_diffex = TRUE
} else {
    show_background_diffex = FALSE
    show_methods_diffex = FALSE
    show_results_diffex = FALSE
    show_appendix_diffex = FALSE
}

if('methods' %in% custom_choices){
    show_methods_custom = TRUE
} else {
    show_methods_custom = FALSE
}
if('results' %in% custom_choices){
    show_results_custom = TRUE
} else {
    show_results_custom = FALSE
}
if('appendix' %in% custom_choices){
    show_appendix_custom = TRUE
} else {
    show_appendix_custom = FALSE
}



show_follow_up = snakemake@config[['report_info']][['include_follow_up']]


```

*Contact:*

* `r analyst_name` <`r analyst_email`>

<!---
Background
--->


# Background

```{r background_biology_section, child = 'child_background_biology.Rmd', eval = show_background_biology}
```

```{r background_samples_section, child = 'child_background_samples.Rmd', eval = show_background_samples}
```

<!---
Methods
--->


# Methods

```{r methods_wetlab_section, child = 'child_methods_wetlab.Rmd', eval = show_methods_wetlab}
```


## Pipeline

We use Snakemake [@Koster2012] to manage the bioinformatics workflow in a reproducible manner.

```{r methods_qc_sequencing_section, child = 'child_methods_qc_sequencing.Rmd', eval = show_methods_qc_sequencing}
```

```{r methods_diffex_section, child = 'child_methods_diffex.Rmd', eval = show_methods_diffex}
```

```{r methods_custom_section, child = 'child_methods_custom.Rmd', eval = show_methods_custom}
```

<!---
Results
--->


# Results

```{r results, child = 'child_results_qc.Rmd', eval = show_results_qc}
```

```{r results, child = 'child_results_diffex.Rmd', eval = show_results_diffex}
```

```{r results_custom_section, child = 'child_results_custom.Rmd', eval = show_results_custom}
```


```{r follow_up, child = 'child_follow_up.Rmd', eval = show_follow_up}
```


# Acknowledgement for Publication

`r acknowledgement_text`

<!---
Appendices
-->

<!-- # Appendices -->

<!-- ```{r appendix_qc_section, child = 'child_appendix_qc.Rmd', eval = show_appendix_qc} -->
<!-- ``` -->

<!-- ```{r appendix_diffex_section, child = 'child_appendix_diffex.Rmd', eval = show_appendix_diffex} -->
<!-- ``` -->

<!-- ```{r appendix_custom_section, child = 'child_appendix_custom.Rmd', eval = show_appendix_custom} -->
<!-- ``` -->

# References
